### PR TITLE
[CA-1327] Github Actions

### PR DESCRIPTION
**Pull request by s-rubenstein**
Ticket: <Link to Jira ticket>
<Put notes here to help reviewer understand this PR>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

**Comment by s-rubenstein** [comment]
I will make the Travis CI piece not required, and make the GHA required before merging, just that is a repo-wide setting so wanted to wait for approvals before futzing with it.

**Comment by andy7i** [comment]
could you add a bit more details about the PR: any differences between travis vs GHA that we should know about? (also, I think we're required to have a ticket # for each PR now)

**Comment by s-rubenstein** [comment]
Yes, thank you. Forgot to update the title after requesting review.

The big differences between GHA and Travis are:

1. There's no separate script for publishing the client library, instead we have some branching logic in the actions file.
2. Test logic and client logic have been separated into two jobs for clarity
3. Artifactory login information is stored in GitHub secrets rather than in Travis env variables.

File: .github/workflows/build.yml
Comment (marctalbott) [inline .github/workflows/build.yml:12]: In Travis we had to specify the Scala version, does GHA know to just default to what Sam uses?

File: .github/workflows/build.yml
Comment (marctalbott) [inline .github/workflows/build.yml]: nit: this runs the unit tests and generates the coverage report

File: .github/workflows/build.yml
Comment (marctalbott) [inline .github/workflows/build.yml]: nit: these aren't integration tests, this command just runs the tests tagged with `SchemaInit`. Maybe rename to 'Run JndiSchemaInit tests' or something like that

File: .github/workflows/build.yml
Comment (marctalbott) [inline .github/workflows/build.yml:106]: do we need to worry about the pull request part of the conditional in the publishJavaClient script?

File: .github/workflows/build.yml
Comment (marctalbott) [inline .github/workflows/build.yml]: nit: the ids are swapped, this one isn't a snapshot

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:12]: I believe so? If not, it is dictated by the coursier/cache-action, which caches at least sbt.

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml]: Sure, can rename.

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:106]: That script goes away with this. Since it had travis env variables anyway.

Comment (marctalbott) [reply .github/workflows/build.yml:106]: Yeah, I was just concerned that the changed conditional would make it so we aren't publishing snapshots in the correct situation. But the `develop` branch seems like the most important part to me, so I guess the pull request part was irrelevant?

**Review by marctalbott — APPROVED** [review]

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:106]: Ah! No, sorry! That is covered by this line https://github.com/broadinstitute/sam/pull/521/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R4

Comment (s-rubenstein) [reply .github/workflows/build.yml:106]: It will only run on push or pull request, due to the nature of the workflow. It will only ever be push on dev.

Comment (s-rubenstein) [reply .github/workflows/build.yml:106]: Although… I guess if we pull request from dev to a branch, it would publish the dev snapshot? But I think that is okay, since it will only ever release a main version from up to date dev code.

File: .github/workflows/build.yml
Comment (andy7i) [inline .github/workflows/build.yml:47]: do we know why travis had this:

```
- docker restart opendj
- sleep 40
```

and whether we need to do something similar (whether restarting or waiting 40 seconds for opendj to start up)? I see that it passed in this test but just checking in case it could cause a race condition and test flakiness.

File: .github/workflows/build.yml
Comment (andy7i) [inline .github/workflows/build.yml:19]: any reason for swapping to port 5432 from 15342?

Comment (andy7i) [reply .github/workflows/build.yml:19]: oh maybe this was why travis used 15432 instead of postgres's default 5432 port:
> Port 15432 is used here for the local machine, rather than using 5432, in case an instance of PostgreSQL was also running on the local machine. If an instance of PostgreSQL was running on the local machine and the same port was used, setting up the connection would fail as the port would already be in use.
https://www.openshift.com/blog/openshift-connecting-database-using-port-forwarding

this probably isn't an issue for us here though

File: .travis.yml
Comment (andy7i) [inline .travis.yml:32]: i _think_ we can remove this dependency https://github.com/broadinstitute/sam/blob/c34ff18ddb109aa15e4f4cade8377f582414c86c/project/plugins.sbt#L5

File: .github/workflows/build.yml
Comment (andy7i) [inline .github/workflows/build.yml]:
```suggestion
      - name: Publish java client for merge to develop branch
```

File: .github/workflows/build.yml
Comment (andy7i) [inline .github/workflows/build.yml:7]: not sure why, but the tests for this PR seem to be running duplicate tests for each(?) commit, one for push and one for pull_request? maybe we really just need `push` if that runs on all pushed commits. not a big deal if we like it this way

```
Scala tests with coverage / test (pull_request) Successful in 6m
Scala tests with coverage / test (push) Successful in 7m

Scala tests with coverage / generateAndPublishClient (pull_request) Successful in 1m
Scala tests with coverage / generateAndPublishClient (push) Successful in 1m
```

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:47]: I am not sure. Since it is unit tests, I am not too worried? Since otherwise it would be a sign of data leakage being an issue in our testing, which I would want to solve anyway.

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:19]: This ^. Seemed like not an issue for GHA

File: .github/workflows/build.yml
Comment (s-rubenstein) [reply .github/workflows/build.yml:7]: IDK, but it mirrors what travis did (with two identical test runs), and I am trying to keep it as similar as possible for the moment. I think we can make later improvements down the line if it makes sense to.

**Review by andy7i — APPROVED** [review]
👍 thanks for making these changes / answering questions!

**Pull request merged**
